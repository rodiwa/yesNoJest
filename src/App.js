import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import ask from './api'

class App extends Component {
  constructor(props) {
    super(props)
    this.getAnswer = this.getAnswer.bind(this)
    this.state = {  
      answer: null
    }
  }

  getAnswer = () => ask().then(answer => this.setState({ answer }))

  render() {
    return (
      <div>
        <div className='question'>
          <input type='text' placeholder='What is your question?' />
          <button
            type='submit'
            onClick={ this.getAnswer }>
            Ask God!
          </button>
        </div>

        { this.state.answer && <div className='answer'>
          <h1>{ this.state.answer.answer }</h1>
          <img src={this.state.answer.image} />
        </div> }
      </div>
    );
  }
}

export default App;
