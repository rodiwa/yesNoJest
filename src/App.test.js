import React from 'react';
import { shallow } from 'enzyme'

import App from './App';
import ask from './api'

jest.mock('./api')

describe('snapshots', () => {
  it('should render the app correctly', () => {
    const wrapper = shallow(<App />)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('app UI should display correctly', () => {

  it('should display text input for asking question', () => {
    const wrapper = shallow(<App />)
    expect(wrapper.find('input[type="text"]').length).toBe(1)
  })

  it('should display button input to ask question', () => {
    const wrapper = shallow(<App />)
    expect(wrapper.find('button').length).toBe(1)
  })

  it('should display answer with text and image', () => {
    const wrapper = shallow(<App />)
    wrapper.find('button').simulate('click')

    return Promise.resolve()
      .then(() => {
        wrapper.update()
    
        const answer = wrapper.find('.answer')
        expect(answer.find('h1').text()).toBe('no')
        expect(answer.find('img').prop('src')).toBe('https://yesno.wtf/assets/no/0-b6d3e555af2c09094def76cf2fbddf46.gif')
      })
  })

  it('should not display any answer by default', () => {
    const wrapper = shallow(<App />)
    const answer = wrapper.find('.answer')
    expect(answer.length).toBe(0)
  })
})

describe('api tests', () => {
  afterEach(() => {
    jest.resetAllMocks()
  })
  
  it.skip('should call yesNo api', () => {
    const fetchSpy = jest.spyOn(global, 'fetch')
      .mockImplementation(() => Promise.resolve({
        json: () => {}
      }))

    return ask()
      .then(() => {
        expect(fetchSpy).toHaveBeenCalledWith('https://yesno.wtf/api/')
      })
  })

  it('should return yesNo response in JSON format', () => {
    return ask()
      .then(response => {
        expect(response).toHaveProperty('answer')
        expect(response).toHaveProperty('image')
      })
  })
  
})
