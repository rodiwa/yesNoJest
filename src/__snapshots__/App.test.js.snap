// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should render the app correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="question"
>
          <input
                    placeholder="What is your question?"
                    type="text"
          />
          <button
                    onClick={[Function]}
                    type="submit"
          >
                    Ask God!
          </button>
</div>,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              placeholder="What is your question?"
              type="text"
/>,
            <button
              onClick={[Function]}
              type="submit"
>
              Ask God!
</button>,
          ],
          "className": "question",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "placeholder": "What is your question?",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Ask God!",
              "onClick": [Function],
              "type": "submit",
            },
            "ref": null,
            "rendered": "Ask God!",
            "type": "button",
          },
        ],
        "type": "div",
      },
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="question"
>
            <input
                        placeholder="What is your question?"
                        type="text"
            />
            <button
                        onClick={[Function]}
                        type="submit"
            >
                        Ask God!
            </button>
</div>,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                placeholder="What is your question?"
                type="text"
/>,
              <button
                onClick={[Function]}
                type="submit"
>
                Ask God!
</button>,
            ],
            "className": "question",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "placeholder": "What is your question?",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Ask God!",
                "onClick": [Function],
                "type": "submit",
              },
              "ref": null,
              "rendered": "Ask God!",
              "type": "button",
            },
          ],
          "type": "div",
        },
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
